library/rsa.c

```diff
@@ -1523,6 +1523,112 @@ int mbedtls_rsa_rsassa_pss_sign( mbedtls_rsa_context *ctx,
  * Implementation of the PKCS#1 v2.1 RSASSA-PKCS1-V1_5-SIGN function
  */
 
+/* Encode a hash into a DigestInfo structure as specified by PKCS#1
+ * (RFC 8017, EMSA-PKCS1-v1_5-ENCODE step 2).
+ * Write to the left of p and set *p to the leftmost byte written. */
+static int rsa_emsa_pkcs1_v15_encode_digestinfo( unsigned char **p,
+                                                 unsigned char *start,
+                                                 mbedtls_md_type_t md_alg,
+                                                 const unsigned char *hash,
+                                                 size_t hashlen )
+{
+    const mbedtls_md_info_t *md_info;
+    const char *oid;
+    size_t oid_size;
+
+    if( md_alg == MBEDTLS_MD_NONE )
+    {
+        if( *p < start + hashlen )
+            return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
+        *p -= hashlen;
+        memcpy( *p, hash, hashlen );
+        return( 0 );
+    }
+
+    md_info = mbedtls_md_info_from_type( md_alg );
+    if( md_info == NULL )
+        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
+    if( hashlen == 0 )
+        hashlen = mbedtls_md_get_size( md_info );
+    else if ( hashlen != mbedtls_md_get_size( md_info ) )
+        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
+    if( mbedtls_oid_get_oid_by_md( md_alg, &oid, &oid_size ) != 0 )
+        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
+
+    /* Double-check that 8 + hashlen + oid_size can be used as a
+     * 1-byte ASN.1 length encoding and that there's no overflow. */
+    if( 8 + hashlen + oid_size  >= 0x80         ||
+        10 + hashlen            <  hashlen      ||
+        10 + hashlen + oid_size <  10 + hashlen )
+        return( MBEDTLS_ERR_RSA_BAD_INPUT_DATA );
+
+    /*
+     * Static bounds check:
+     * - Need 10 bytes for five tag-length pairs.
+     *   (Insist on 1-byte length encodings to protect against variants of
+     *    Bleichenbacher's forgery attack against lax PKCS#1v1.5 verification)
+     * - Need hashlen bytes for hash
+     * - Need oid_size bytes for hash alg OID.
+     */
+    if( *p < start + 10 + oid_size + hashlen )
```

Comment:
As above, this should be `if( *p - start < 10 + oid_size + hashlen )` instead.

Reply:
Done with a slight modification.